File: 69.sqrt-x.py
#
# @lc app=leetcode id=69 lang=python
#
# [69] Sqrt(x)
#

# @lc code=start
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x <= 1:
            return x
        l, h = 1, x
        while l <= h:
            mid = l + int((h-l) / 2)
            sqrt = int(x / mid)
            if sqrt == mid:
                return mid
            elif mid > sqrt:
                h = mid - 1
            else:
                l = mid + 1
        return h
# @lc code=end

File: 69.sqrt-x.cpp
/*
 * @lc app=leetcode id=69 lang=cpp
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution {
public:
    int mySqrt(int x) {
        if (x<=1)
            return x;

        int l = 1, h = x;
        while (l<=h) {
            int mid = l + (h - l) / 2;
            int sqrt = x / mid;
            if (sqrt == mid)
                return mid;
            else if (mid > sqrt)
                h = mid - 1;
            else
                l = mid + 1;
        }
        return h;
    }
};
// @lc code=end

